grr/server/grr_response_server/databases/spanner_paths.py

```diff
@@ -0,0 +1,636 @@
+        int_path_type = int(path_info.path_type)
+        path = EncodePathComponents(path_info.components)
+
+        path_row = [client_id, int_path_type, path,
+                    spanner_lib.COMMIT_TIMESTAMP,
+                    path_info.directory,
+                    len(path_info.components)
+        ]
+
+        if path_info.HasField("stat_entry"):
+          path_columns.append("LastFileStatTime")
+          path_row.append(spanner_lib.COMMIT_TIMESTAMP)
+
+          file_stat_row = [client_id, int_path_type, path,
+                           spanner_lib.COMMIT_TIMESTAMP,
+                           path_info.stat_entry
+          ]
+        else:
+          file_stat_row = None
+
+        if path_info.HasField("hash_entry"):
+          path_columns.append("LastFileHashTime")
+          path_row.append(spanner_lib.COMMIT_TIMESTAMP)
+
+          file_hash_row = [client_id, int_path_type, path,
+                           spanner_lib.COMMIT_TIMESTAMP,
+                          path_info.hash_entry,
+          ]
+        else:
+          file_hash_row = None
+
+        mut.insert_or_update(table="Paths", columns=path_columns, values=[path_row])
+        if file_stat_row is not None:
+          mut.insert(table="PathFileStats", columns=file_stat_columns, values=[file_stat_row])
+        if file_hash_row is not None:
+          mut.insert(table="PathFileHashes", columns=file_hash_columns, values=[file_hash_row])
+
+        for path_info_ancestor in models_paths.GetAncestorPathInfos(path_info):
+          components = tuple(path_info_ancestor.components)
+          ancestors.add((path_info.path_type, components))
+
+      path_columns = ["ClientId", "Type", "Path", "CreationTime", "IsDir", "Depth"]
+      for path_type, components in ancestors:
+        path_row = [client_id, int(path_type), EncodePathComponents(components),
+                    spanner_lib.COMMIT_TIMESTAMP, True, len(components),
+        ]
+        mut.insert_or_update(table="Paths", columns=path_columns, values=[path_row])
+
+    try:
+      self.db.Mutate(Mutation, txn_tag="WritePathInfos")
+    except NotFound as error:
+      if "Parent row for row [" in str(error):
+        raise abstract_db.UnknownClientError(client_id) from error
+      else:
+        raise
+
+  @db_utils.CallLogged
+  @db_utils.CallAccounted
+  def ReadPathInfos(
+      self,
+      client_id: str,
+      path_type: objects_pb2.PathInfo.PathType,
+      components_list: Collection[Sequence[str]],
+  ) -> dict[tuple[str, ...], Optional[objects_pb2.PathInfo]]:
+    """Retrieves path info records for given paths."""
+    # Early return to avoid issues with unnesting empty path list in the query.
+    if not components_list:
+      return {}
+
+    results = {tuple(components): None for components in components_list}
+
+    query = """
+    SELECT p.Path, p.CreationTime, p.IsDir,
+           ps.CreationTime, ps.Stat,
+           ph.CreationTime, ph.FileHash,
```

Comment:
```suggestion
           ph.CreationTime, ph.FileHash
```

Remove the trailing comma here